# Исследование надежности заемщиков


Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [34]:
# создадим словарь с парами: количество детей - доля невозвращенных кредитов
ctd = {}
for children in sorted(data['children'].unique()):
    ctd[children] = data.loc[data['children'] == children]['debt'].mean()
ctd

{0: 0.07543822297920659,
 1: 0.09234608985024959,
 2: 0.09454191033138401,
 3: 0.08181818181818182,
 4: 0.0975609756097561,
 5: 0.0}

In [35]:
# приведем словарь в более наглядный вид
columns = ['number_of_children', 'debt_percentage']
children_to_debt = pd.DataFrame(ctd.items(), columns=columns)
children_to_debt['number_of_loaners'] = data.groupby('children')['debt'].count()
children_to_debt['number_of_debtors'] = data.groupby('children')['debt'].sum()
children_to_debt.style.format({'debt_percentage' : '{:.2%}'.format})

**Вывод:** 

Заемщики без детей лучше всего отдают кредиты, доля должников здесь составляет 7,54%; заемщики с одним или двумя детьми отдают кредиты менее охотно - 9,23% и 9,45% должников соответственно. Выборка для заемщиков с тремя и больше детьми слишком мала, данных недостаточно для корректных выводов по данным категориям.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [24]:
# по аналогии с предыдущим пунктом создадим словарь с парами: семейное положение - доля невозвращенных кредитов
std = {}
for status in sorted(data['family_status'].unique()):
    std[status] = data.loc[data['family_status'] == status]['debt'].mean()
std

{'Не женат / не замужем': 0.09763948497854077,
 'в разводе': 0.07064760302775441,
 'вдовец / вдова': 0.06624605678233439,
 'гражданский брак': 0.09313014029995162,
 'женат / замужем': 0.07560557866405676}

In [50]:
# создадим наглядную таблицу
columns = ['family_status', 'debt_percentage']
status_to_debt = pd.DataFrame(std.items(), columns=columns)
status_to_debt['number_of_loaners'] = data.groupby('family_status')['debt'].count().values
status_to_debt['number_of_debtors'] = data.groupby('family_status')['debt'].sum().values
status_to_debt.style.format({'debt_percentage' : '{:.2%}'.format})

**Вывод:** 

Среди неженатых или состоящих в гражданском браке заемщиков самый высокий процент должников - 9,76% и 9,31% соответственно. Заемщики, состоящие в браке или разведенные относятся к кредитам более ответственно - 7,56% и 7,06% должников соответственно. Для вдов и вдовцов выборка слишком мала, чтобы делать корректные выводы. 

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [26]:
# здесь словарь с парами: категория уровня дохода - доля невозвращенных кредитов
itd = {}
for income in sorted(data['total_income_category'].unique()):
    itd[income] = data.loc[data['total_income_category'] == income]['debt'].mean()
itd

{'A': 0.08,
 'B': 0.07060231352213801,
 'C': 0.08498209911437724,
 'D': 0.06017191977077364,
 'E': 0.09090909090909091}

In [51]:
# таблица
columns = ['total_income_category', 'debt_percentage']
income_to_debt = pd.DataFrame(itd.items(), columns=columns)
income_to_debt['number_of_loaners'] = data.groupby('total_income_category')['debt'].count().values
income_to_debt['number_of_debtors'] = data.groupby('total_income_category')['debt'].sum().values
income_to_debt.style.format({'debt_percentage' : '{:.2%}'.format})

**Вывод:** 

Самый низкий процент должников среди тех заемщиков, чей доход составляет от 200 000 до 1 000 000 - 7,06% должников, самый высокий - у тех, чей доход от 50 000 до 200 000 - 8,50% должников. Для всех остальных мы имеем слишком маленькую выборку, и сравнивать некорректно.


#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [28]:
# словарь с парами: категория цели кредита - доля невозвращенных кредитов
ptd = {}
for purpose in sorted(data['purpose_category'].unique()):
    ptd[purpose] = data.loc[data['purpose_category'] == purpose]['debt'].mean()
ptd

{'операции с автомобилем': 0.09347978499649451,
 'операции с недвижимостью': 0.07255139056831923,
 'получение образования': 0.09252758274824474,
 'проведение свадьбы': 0.07911802853437094}

In [56]:
# таблица
columns = ['purpose_category', 'debt_percentage']
purpose_to_debt = pd.DataFrame(ptd.items(), columns=columns)
purpose_to_debt['number_of_loaners'] = data.groupby('purpose_category')['debt'].count().values
purpose_to_debt['number_of_debtors'] = data.groupby('purpose_category')['debt'].sum().values
purpose_to_debt.style.format({'debt_percentage' : '{:.2%}'.format})

**Вывод:** 

Кредиты, взятые на автомобиль или образование (9,35% и 9,25% должников соответственно) возвращают реже, чем кредиты, взятые на недвижимость или свадьбу (7.26% и 7,91% должников). И это справедливо, ведь должникам по ипотеке грозит изъятие ипотечной недвидимости.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 

Все пропуски находятся в двух столбцах - с данными о трудовом стаже и доходах, причем в одной строке оба значения пропущены. Это может быть связано с тем, что заемщик не предоставил эту информацию.

Возможно, данные были утеряны из-за технического сбоя в системе сбора информации. 

Возможно, [ДАННЫЕ УДАЛЕНЫ] были утеряны в ходе ████████████ ████ или из-за ██████ ███ ████████.(Очень захотелось пошутить, извини, если неуместно)

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 

Медианные значения используют для получения более объективной информации, если в выборке присутствуют выдаюшиеся значения

### Шаг 4: общий вывод.

В этом проекте я изучила данные, нашла в них пропуски (порядка 2000 строк с пропусками значений в столбце с данными о трудовом стаже и данными о доходах) и заполнила их медианными значениями.

Нашла аномалии, такие как отрицательные значения в столбце с данными о трудовом стаже, а также нереалистичные данные о количестве детей; первую аномалию исправила, изменив значения на положительные, вторую - удалив строки с нереалистичными значениями.

Нашла и обработала неявные дубликаты в столбце с данными об образовании, приведя все значения к нижнему регистру, после чего удалила явные дубликаты (71 строка).

Создала два новых столбца с категориями для двух столбцов: с данными о доходе и с целями взятия кредита.

Используя обработанный датафрейм, смогла найти зависимости между невозвращенными кредитами и количеством детей, семейным положением, уровнем дохода и целью кредита.

Заемщики без детей вероятнее вернут кредит в срок, чем заемщики с ребенком или двумя, по статистике здесь 7,54% против 9,23% и 9,45% соответственно.

Заемщики, состоящие в браке или разведенные, вероятнее вернут кредит, чем не состоящие или состоящие в гражданском браке, по статистике: 7,56% и 7,06% против 9,76% и 9,31%.

Заемщики, имеющие доход от 200 000 до 1 000 000 вероятнее вернут кредит, чем имеющие доход от 50 000 до 200 000, по статистике 7,06% против 8,50%.

Заемщики, планирующие свадьбу или покупку недвижимости вероятнее вернут кредит, чем планирующие покупку автомобия или получение образования, по статистике 7,91% и 7,26% против 9,35% и 9,25%.

Однако, в данных были обнаружены пропуски (в порядка 2 000 строк), которые могли повлиять на результаты исследования. Этим пропускам, возможно, стал причиной сбой в системе сбора информации, это стоит проверить.

Также были обнаружены категории, для которых данных недостаточно для корректных выводов, такие как: заемщики с тремя и более детьми, вдовы/вдовцы, заемщики с доходом менее 50 000 или более 1 000 000. В таких случаях стоит уделить заемщикам особое внимание, а не полагаться на статистику.